File: Java/MinCollector.java
public class MinCollector {
    private final int threadNum;
    private int threadCount = 0;
    private int globalMin = Integer.MAX_VALUE;
    private int globalMinIndex = 0;

    public MinCollector(int threadNum) {
        this.threadNum = threadNum;
    }

    public synchronized void collect(int min, int index) {
        if (min < globalMin) {
            globalMin = min;
            globalMinIndex = index;
        }
    }

    public synchronized void incThreadCount() {
        threadCount++;
        if (threadCount == threadNum) {
            notify();
        }
    }

    public synchronized int[] getResult() {
        while (threadCount < threadNum) {
            try {
                wait();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return new int[]{globalMin, globalMinIndex};
    }
}

File: C#/MinCollector.cs
using System.Threading;

namespace LB1
{
    public class MinCollector
    {
        private int globalMin = int.MaxValue;
        private int globalMinIndex = -1;
        private int finishedThreads = 0;
        private readonly int totalThreads;
        private readonly object lockObj = new object();

        public MinCollector(int threadCount)
        {
            totalThreads = threadCount;
        }

        public void Collect(int localMin, int localMinIndex)
        {
            lock (lockObj)
            {
                if (localMin < globalMin)
                {
                    globalMin = localMin;
                    globalMinIndex = localMinIndex;
                }

                finishedThreads++;

                if (finishedThreads == totalThreads)
                {
                    Monitor.Pulse(lockObj); // будимо потік, який чекає
                }
            }
        }

        public (int, int) GetResult()
        {
            lock (lockObj)
            {
                while (finishedThreads < totalThreads)
                {
                    Monitor.Wait(lockObj); // чекаємо завершення всіх потоків
                }

                return (globalMin, globalMinIndex);
            }
        }
    }
}
